Csweaver/addbackend

csweaver opened the pull request.

ttung commented on server/app/scanpy_engine/scanpy_engine.py:
```diff
@@ -0,0 +1,183 @@
+    def genes(self):
+        return self.data.var.index.tolist()
+
+    def filter_cells(self, filter):
```
To better understand this, I think it would be helpful to know what filter does and behaves.

csweaver replied:
Here is how filters work: 
For categorical metadata keys filter based on key=value 
For continuous metadata keys filter by key=min,max
Either value can be replaced by a *. To have only a minimum value key=min,* To have only a maximum value key=*,max 

They combine via AND so a cell's metadata would have to match every filter

The results is a matrix with the cells the pass the filter and at this point all the genes

ttung commented on server/app/util/filter.py:
```diff
@@ -0,0 +1,64 @@
+    :raises: ValueError
+    """
+    try:
+        if variable and datatype == "int":
```
doesn't this break down if variable is 0?

csweaver replied:
yes it does, good catch. I'll fix it

csweaver commented on server/app/driver/driver.py:
```diff
@@ -0,0 +1,78 @@
+        """
+        pass
+
+    # Should this return the order of metadata fields as the first value?
```
Ok, good. I'll add that they are returned in the same order in the comments

csweaver commented on server/app/scanpy_engine/scanpy_engine.py:
```diff
@@ -0,0 +1,183 @@
+        return data_schema
+
+    def _set_cell_ids(self):
+        self.data.obs['cxg_cell_id'] = list(range(self.data.obs.shape[0]))
```
oops, I'll fix that

csweaver commented on server/app/driver/driver.py:
```diff
@@ -0,0 +1,78 @@
+        pass
+
+    @abstractmethod
+    def _set_cell_ids(self):
```
Sets an internal id for cells in the dataframe. 

Since I punted on the API v2.0 I think I should remove the references to cellids

csweaver commented on server/app/scanpy_engine/scanpy_engine.py:
```diff
@@ -0,0 +1,183 @@
+    def cells(self):
+        return list(self.data.obs.index)
+
+    def cellids(self, df=None):
```
I'm going to remove cell id references, see above

csweaver commented:
@ttung I've fixed your requests. For the filter bug, I realized there was a bug, but it was even more complex than what you pointed out. I corrected it and added in unit tests related to that file.

csweaver commented:
@ttung Thank you!! Added the filter documentation, fixed the */. issue. Good catch on the cell ids, I had fixed that in the driver but not in the engine. I also fixed a problem with the requirements file. 

I'm going to hold off on merging this at @colinmegill's request until he's ready for it.

Def with add more tests and travis integration!

csweaver commented:
Merged. Unfortunately doing that via the command line does not show that here.
